Validate: get line numbers when calling validate on ignition structs

ajeddeloh opened the pull request:
This makes the converter functions also modify the yaml tree, converting it into a tree that looks like it was parsed from a json source. Validate() can then be used instead of ValidateWithoutSource(), exposing locations of errors.

Depends on https://github.com/coreos/ignition/pull/355

ajeddeloh commented on config/astyaml/astyaml.go:
```diff
@@ -12,79 +12,127 @@
+	if n.Kind != yaml.MappingNode {
+		return ErrNotMappingNode
+	}
+	for i := 0; i < len(n.Children); i += 2 {
```
go-yaml parses a mapping type into an array of interleaved key and value nodes, so even nodes are the keys and odd nodes the values.

ajeddeloh commented on config/types/config.go:
```diff
@@ -49,36 +50,48 @@ type ConfigReference struct {
+			r.Merge(convertReport)
+			if convertReport.IsFatal() {
+				// don't add to the output if invalid
+				continue
```
We do. It gets added to the report that gets returned. It can continue trying to parse of sections of the config and catch multiple errors at once instead of failing at the first one.

ajeddeloh commented on config/astyaml/astyaml.go:
```diff
@@ -12,79 +12,127 @@
+	return n.tag
+}
+
+// ChangeTreeTag changes the value Tag() returns so it
```
Oops, yeah the comment is incomplete. I'll change it to have a pointer receiver and modify the node itself.

ajeddeloh commented:
It does feel a little hacky, but I'm not sure a good way around it. You can take out the parts the modify the tree and just leave the changing of tag to "json" and you retain line numbers for all parts that are identical but lose line number accuracy for the parts that are subtly different (it'll report the line number of the last node that was the same).

dgonyeo commented:
ct [moved to using glide](https://github.com/coreos/container-linux-config-transpiler/pull/63), so you'll need to rebase and redo the vendor work with glide. If you need glide help lmk.

dghubble commented:
Sorry to complicate your PR. You can go get glide github.com/Masterminds/glide, edit `glide.yaml` to declare any new dependencies or pinned versions (if any), and run `make vendor`.

ajeddeloh commented:
Fixed the issue the ChangeTreeTag and rebased.

glide.yaml currently uses my fork of ignition for a needed change (https://github.com/coreos/ignition/pull/355) and once that is merged I can change that back.

dgonyeo commented on config/astyaml/astyaml.go:
```diff
@@ -12,79 +12,128 @@
+		if key.Value == oldKeyName {
+			//key.Value = newKeyName
+			(*n.Children[i]).Value = newKeyName
+			*n.Children[i+1] = newValue.Node
```
So based on your earlier comment, `n.Children` is _guaranteed_ to have an even length? Just want to make sure this won't panic.

ajeddeloh replied:
Yes. See https://github.com/ajeddeloh/yaml/blob/public-node/decode.go#L175 The key/values are always added in pairs.

dgonyeo commented on config/astyaml/astyaml.go:
```diff
@@ -12,79 +12,128 @@
+		key:  n.key,
+		tag:  newTag,
+		Node: n.Node,
+	}*/
```
I take it this comment can be deleted?

dgonyeo commented on config/types/common.go:
```diff
@@ -114,3 +119,41 @@ func getArgs(format, tagName string, e interface{}) []string {
+
+// Get returns the value for key, where key is an int or string and n is a
+// sequence node or mapping node, respectively.
+func get(n validate.AstNode, key interface{}) (validate.AstNode, error) {
```
I'd prefer this to be named something a little more descriptive, since there are other functions here like `getCliArgs` that are entirely unrelated to `get`.

dgonyeo commented on config/types/converter.go:
```diff
@@ -17,20 +17,29 @@ package types
+func ConvertAs2_0(in Config, platform string, ast validate.AstNode) (ignTypes.Config, report.Report) {
+	// convert our tree from having yaml tags to having json tags, so when Validate() is
+	// called on the tree, it can find the keys in the ignition structs (which are denoted
+	// by `json` tags
```
)

ajeddeloh commented:
Added unit tests.

ajeddeloh commented:
Ignition vendoring done.

dgonyeo commented:
can you rebase?

ajeddeloh commented:
rebased
